deviceplugin/usb.go

```diff
@@ -0,0 +1,213 @@
+func (dev *usbDevice) BusPath() (path string) {
+	return fmt.Sprintf(usbDevBus, dev.Bus, dev.BusDevice)
+}
+
+// queryUSBDeviceCharacteristicsByDirectory scans the given directory for information regarding the given USB device,
+// then returns a pointer to a new usbDevice if information is found.
+// Safe to presume that result is set if err is nil.
+func queryUSBDeviceCharacteristicsByDirectory(dir os.DirEntry) (result *usbDevice, err error) {
+	if !dir.IsDir() {
+		// There shouldn't be any raw files in this directory, but just in case.
+		return result, fmt.Errorf("not a directory")
+	}
+	// Try to find the vendor ID file inside this device - this is a good indication that we're dealing with a device, not a bus.
+	vnd, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirVendorIDFile)
+	if err != nil {
+		// We can't read the vendor file for some reason, it probably doesn't exist.
+		return result, err
+	}
+	prd, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirProductIDFile)
+	if err != nil {
+		return result, err
+	}
+
+	// The following two calls shouldn't fail.
+	busRaw, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirBusFile)
+	if err != nil {
+		return result, err
+	}
+	bus := binary.LittleEndian.Uint16(busRaw)
+	busLocRaw, err := os.ReadFile(dir.Name() + "/" + usbDevicesDirBusDevFile)
+	if err != nil {
+		return result, err
+	}
+	busLoc := binary.LittleEndian.Uint16(busLocRaw)
+
+	res := usbDevice{
+		Vendor:    usbID(binary.LittleEndian.Uint16(vnd)),
+		Product:   usbID(binary.LittleEndian.Uint16(prd)),
+		Bus:       bus,
+		BusDevice: busLoc,
+	}
+	return &res, nil
+}
+
+// enumerateUSBDevices rapidly scans the OS system bus for attached USB devices.
+// Pure Go; does not require external linking.
+func enumerateUSBDevices() (specs []usbDevice, err error) {
+	allDevs, err := os.ReadDir(usbDevicesDir)
+	if err != nil {
+		return
+	}
+
+	// Set up a WaitGroup with a buffered channel for results
+	var wg sync.WaitGroup
+	devs := make(chan *usbDevice, len(allDevs))
+
+	// You could also have a shared slice with a mutex guard, but this way is arguably a little more performant.
+	for _, dev := range allDevs {
+		// Copy the loop variable
+		dev := dev
+
+		// Spawn a goroutine to discover the device information
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			result, err := queryUSBDeviceCharacteristicsByDirectory(dev)
+			if err != nil {
+				// do we want to handle errors here?
+				return
+			}
+			// Successful data will get thrown onto the buffered channel for later merging
+			devs <- result
+		}()
+	}
+
+	wg.Wait()
+	close(devs)
+	// Now unwind the buffer into the results array
+	for d := range devs {
+		specs = append(specs, *d)
+	}
+
+	return
+}
+
+// searchUSBDevices returns a subset of the "devices" slice containing only those usbDevices that match the given vendor and product arguments.
+func searchUSBDevices(devices *[]usbDevice, vendor usbID, product usbID) (devs []usbDevice, err error) {
```

Comment:
there's no need to pass the slice by reference using a pointer, since slices are already reference types
```suggestion
func searchUSBDevices(devices []usbDevice, vendor usbID, product usbID) (devs []usbDevice, err error) {
```

Reply:
https://medium.com/swlh/golang-tips-why-pointers-to-slices-are-useful-and-how-ignoring-them-can-lead-to-tricky-bugs-cac90f72e77b I generally tend to prefer passing pointers to mutables purely out of habit to avoid gotchas (especially if something happens later down the line that means we start threading calls to `searchUSBDevices()`, open to further discussion however